File: portal/api/lib/policyReader.ts
/**
 * policy_settings tablosundan yapılandırma değeri okur.
 * owner→global (NULL) fallback zinciri; 5 dakika in-memory cache.
 * DB hatası veya parse hatası → sessizce fallback döner.
 */

import type { SupabaseClient } from '@supabase/supabase-js'

interface CacheEntry {
  rawJson: unknown
  expiry: number
}

const cache = new Map<string, CacheEntry>()
const CACHE_TTL_MS = 5 * 60 * 1000 // 5 dakika

async function fetchRaw(
  supabase: SupabaseClient,
  ownerId: string | null,
  key: string,
): Promise<unknown | undefined> {
  const cacheKey = `${ownerId ?? ''}|${key}`
  const entry = cache.get(cacheKey)
  if (entry && entry.expiry > Date.now()) return entry.rawJson

  try {
    let query = supabase
      .from('policy_settings')
      .select('value')
      .eq('key', key)
      .limit(1)

    if (ownerId) {
      query = query.eq('owner_user_id', ownerId) as typeof query
    } else {
      query = query.is('owner_user_id', null) as typeof query
    }

    const { data, error } = await query
    if (error || !data || data.length === 0) return undefined

    const raw = (data[0] as { value: unknown }).value
    cache.set(cacheKey, { rawJson: raw, expiry: Date.now() + CACHE_TTL_MS })
    return raw
  } catch {
    return undefined
  }
}

/**
 * policy_settings'ten değer okur.
 * owner satırı → global satır → fallback zinciri.
 * Parse hatası veya DB yoksa fallback döner.
 */
export async function getPolicy<T>(
  supabase: SupabaseClient,
  ownerId: string | null | undefined,
  key: string,
  fallback: T,
): Promise<T> {
  const candidates: Array<string | null> = []
  if (ownerId) candidates.push(ownerId)
  candidates.push(null) // global

  for (const owner of candidates) {
    const raw = await fetchRaw(supabase, owner, key)
    if (raw === undefined || raw === null) continue
    // Değer zaten parse edilmiş JSONB değeri (Supabase JS otomatik parse eder)
    try {
      return raw as T
    } catch {
      continue
    }
  }

  return fallback
}

export function invalidatePolicyCache() {
  cache.clear()
}

File: src/AgentArmy.Cli/Infra/PolicyReader.cs
using System.Collections.Concurrent;
using System.Text.Json;

namespace AgentArmy.Cli;

/// <summary>
/// policy_settings tablosundan yapılandırma değeri okur.
/// Önce owner satırı, sonra global (NULL) satır aranır; DB yoksa veya parse hatası olursa fallback döner.
/// 5 dakika in-memory cache; thread-safe.
/// </summary>
public static class PolicyReader
{
    private sealed record CacheEntry(string RawJson, DateTime Expiry);

    // cache key: "ownerId|key" veya "|key" (global)
    private static readonly ConcurrentDictionary<string, CacheEntry> _cache = new();
    private static readonly TimeSpan CacheTtl = TimeSpan.FromMinutes(5);

    /// <summary>
    /// owner→global fallback zinciriyle değer okur.
    /// DB yoksa veya tablo erişim hatası varsa <paramref name="fallback"/> döner.
    /// Parse hatası varsa da <paramref name="fallback"/> döner (sessiz).
    /// </summary>
    public static async Task<T> GetAsync<T>(
        SupabaseWriter? db,
        string? ownerId,
        string key,
        T fallback,
        CancellationToken ct)
    {
        if (db is null) return fallback;

        // Önce owner satırı, sonra global satır dene
        var candidates = new List<string?>();
        if (!string.IsNullOrWhiteSpace(ownerId)) candidates.Add(ownerId);
        candidates.Add(null); // global fallback

        foreach (var owner in candidates)
        {
            var raw = await FetchRawAsync(db, owner, key, ct);
            if (raw is null) continue;

            try
            {
                var parsed = JsonSerializer.Deserialize<T>(raw);
                return parsed is null ? fallback : parsed;
            }
            catch
            {
                // Parse hatası: bu owner için sessizce geç, global'ı dene
                continue;
            }
        }

        return fallback;
    }

    private static async Task<string?> FetchRawAsync(
        SupabaseWriter db,
        string? ownerId,
        string key,
        CancellationToken ct)
    {
        var cacheKey = $"{ownerId ?? string.Empty}|{key}";
        if (_cache.TryGetValue(cacheKey, out var entry) && entry.Expiry > DateTime.UtcNow)
            return entry.RawJson;

        try
        {
            // SupabaseWriter.SelectAsync ile: policy_settings?select=value&key=eq.<key>&owner_user_id=eq.<id>
            // owner IS NULL için özel filtre gerekiyor — raw RPC kullanıyoruz.
            // Alternatif: ayrı bir okuma RPC'si; ama SupabaseWriter.SelectAsync parametreleri destekliyorsa onu kullan.
            // Gerçek implementasyon: REST GET ile filtre.
            var filter = ownerId is not null
                ? $"key=eq.{Uri.EscapeDataString(key)}&owner_user_id=eq.{Uri.EscapeDataString(ownerId)}&select=value&limit=1"
                : $"key=eq.{Uri.EscapeDataString(key)}&owner_user_id=is.null&select=value&limit=1";

            // SelectAsync bir JSON array döner: [{"value": ...}]
            var result = await db.SelectAsync("policy_settings", filter, ct);
            if (result.ValueKind != JsonValueKind.Array) return null;
            var arr = result.EnumerateArray().ToList();
            if (arr.Count == 0) return null;

            var row = arr[0];
            if (row.ValueKind != JsonValueKind.Object) return null;
            if (!row.TryGetProperty("value", out var valEl)) return null;

            var raw = valEl.GetRawText();
            _cache[cacheKey] = new CacheEntry(raw, DateTime.UtcNow.Add(CacheTtl));
            return raw;
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[PolicyReader] policy_settings okunamadı (key={key}, owner={ownerId}): {ex.Message}");
            return null;
        }
    }

    /// <summary>Test veya ayar değişikliği için cache'i temizle.</summary>
    public static void InvalidateCache() => _cache.Clear();
}
